pgjdbc/src/main/java/org/postgresql/util/PasswordUtil.java

```diff
@@ -0,0 +1,189 @@
+   * Generate the encoded text representation of the given password for
+   * SCRAM-SHA-256 authentication. The return value of this method is the literal
+   * text that may be used when creating or modifying a user with the given
+   * password without the surrounding single quotes.
+   *
+   * @param password   The plain text of the user's password
+   * @param iterations The number of iterations of the hashing algorithm to
+   *                   perform
+   * @param salt       The random salt value
+   * @return T text representation of the password encrypted for SCRAM-=SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password, int iterations, byte[] salt) {
+    if (iterations <= 0) {
+      throw new IllegalArgumentException("iterations must be greater than zero");
+    }
+    if (salt.length == 0) {
+      throw new IllegalArgumentException("salt length must be greater than zero");
+    }
+    String passwordText = String.valueOf(password);
+    byte[] saltedPassword = ScramFunctions.saltedPassword(ScramMechanisms.SCRAM_SHA_256,
+        StringPreparations.SASL_PREPARATION, passwordText, salt, iterations);
+    byte[] clientKey = ScramFunctions.clientKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+    byte[] storedKey = ScramFunctions.storedKey(ScramMechanisms.SCRAM_SHA_256, clientKey);
+    byte[] serverKey = ScramFunctions.serverKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+
+    return "SCRAM-SHA-256" //
+      + "$" + iterations //
+      + ":" + Base64.toBase64String(salt) //
+      + "$" + Base64.toBase64String(storedKey) //
+      + ":" + Base64.toBase64String(serverKey);
+  }
+
+  /**
+   * Encode the given password for SCRAM-SHA-256 authentication using the default
+   * iteration count and a random salt.
+   *
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for SCRAM-=SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password) {
+    SecureRandom rng = getSecureRandom();
+    byte[] salt = rng.generateSeed(DEFAULT_SALT_LENGTH);
+    return encodeScramSha256(password, DEFAULT_ITERATIONS, salt);
+  }
+
+  /**
+   * Encode the the given password for use with md5 authentication. The PostgreSQL
+   * server uses the username as the per-user salt so that must also be provided.
+   * The return value of this method is the literal text that may be used when
+   * creating or modifying a user with the given password without the surrounding
+   * single quotes.
+   *
+   * @param user     The username of the database user
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for md5
+   *         authentication.
+   */
+  public static String encodeMd5(String user, char[] password) {
+    byte[] userBytes = user.getBytes(StandardCharsets.UTF_8);
+    String passwordText = String.valueOf(password);
```

Comment:
WDYT of `Arrays.fill(password, ' ')` as soon as the password was consumed? (e.g. after conversion no string)

A slightly better approach would probably be `CharBuffer.wrap(password)`, then encode it to byte array with Charset.encode, then zero out both arrays.

Reply:
I don't think it's the job of the encoder to wipe out the caller's data structures. The caller can choose to do that if necessary.

That would make more sense if the function signature was something like:

```java
public static String encodeMd5(String user, Supplier<char[]> passwordSource) { 
  char[] password = passwordSource.get();
  // do stuff
  Arrays.fill(password, ' '); // clear password array
}
```

In that situation we "own" the array and it'd be our responsibility for clearing it.

I'm of two minds here ... partly think we should go that route as it's the most flexible. And partly think nobody is going to care about this and we should just make `String encode(String user, String password)` overloads as that's what 99% of people are going to use.

Reply:
There are two points of using `char[]` for passwords:
1) The API that uses password can erase the password shortly after use so it minimizes the time window when the password can be grabbed from the memory dump
2) There are much less chances of logging the passwords as `char[].toString()` would not dump contents.

In this case, as soon as you compute MD5, you can erase the `char[]` array so the password is not accidentally leaked.

Reply:
I know why it's being done. I'm the one that suggested it in Dave's original PR!

I'm saying the reality of how people program in Java lends itself to this being a moot point because the caller is going to have the password in a String and simply convert it to a char[] at the call site.

Regardless, destroying the contents of an argument (like we'd be doing here without changing the method signature) isn't the same as destroying the contents of the return value of a function (e.g. what AuthenticationPluginManager does).

In this specific case it'd be weird because it'd break the most common use case of this function:

```java
char[] password = generateNewPassword();
conn.alterUserPassword(user, password);
saveNewPasswordSomewhere(password);
```

The user owns the char[] so the user is responsible for clearing it.

Reply:
If the user wants saving the password, they should clone the array